src/typecheck.rs

```diff
@@ -1,11 +1,1384 @@
-//! The type checker
-
 use crate::syntax::*;
+use thiserror::Error;
+
+/// Errors that can occur during type checking
+#[derive(Error, Debug)]
+pub enum TypeError {
+    /// Type mismatch between expected and actual types
+    #[error("Type mismatch: expected {expected:?}, found {actual:?}")]
+    TypeMismatch {
+        /// The type that was expected
+        expected: Type,
+        /// The type that was actually found
+        actual: Type,
+    },
+    /// Variable not found in the environment
+    #[error("Undefined variable: '{0}'")]
+    UndefinedVariable(Id),
+    /// Function not found
+    #[error("Undefined function: '{0}'")]
+    UndefinedFunction(Id),
+    /// Symbol not in alphabet
+    #[error("Symbol '{0}' is not in the alphabet")]
+    SymbolNotInAlphabet(char),
+    /// Wrong number of arguments
+    #[error("Wrong number of arguments: expected {expected}, actual {actual}")]
+    WrongNumberOfArguments {
```

Comment:
ditto (removing the whole thing is fine, because this is part of an error for function types mismatching)